File: DSA/VedantChouhan_590024563/Day-17/q2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct Node {
    int data;
    struct Node *next;
};

struct Node* createNode(int data) {
    struct Node *newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

bool hasCycle(struct Node *head) {

    struct Node *slow = head;
    struct Node *fast = head;

    while (fast != NULL && fast->next != NULL) {

        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast)
            return true;
    }

    return false;
}

int main() {

    int arr[] = {3, 2, 0, -4};
    int n = sizeof(arr) / sizeof(arr[0]);

    struct Node *head = NULL;
    struct Node *tail = NULL;
    struct Node *nodes[100];

    // Create linked list
    for (int i = 0; i < n; i++) {

        struct Node *newNode = createNode(arr[i]);
        nodes[i] = newNode;

        if (head == NULL) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
    }

    // pos = 1 (last node points to node with value 2)
    int pos = 1;

    if (pos != -1)
        tail->next = nodes[pos];

    if (hasCycle(head))
        printf("true\n");
    else
        printf("false\n");

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day17/Q2.java
import java.util.*;

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class Main {

    // Function to detect cycle
    public static boolean hasCycle(Node head) {

        Node slow = head;
        Node fast = head;

        while (fast != null && fast.next != null) {

            slow = slow.next;          // Move one step
            fast = fast.next.next;     // Move two steps

            if (slow == fast) {
                return true;
            }
        }

        return false;
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        // Number of nodes
        int n = sc.nextInt();

        // First node
        Node head = new Node(sc.nextInt());
        Node temp = head;

        // Store all nodes in an array
        Node[] nodes = new Node[n];
        nodes[0] = head;

        // Create remaining nodes
        for (int i = 1; i < n; i++) {
            Node newNode = new Node(sc.nextInt());
            temp.next = newNode;
            temp = newNode;
            nodes[i] = newNode;
        }

        // Position where last node points
        int pos = sc.nextInt();

        // Create cycle if pos != -1
        if (pos != -1) {
            temp.next = nodes[pos];
        }

        // Print result
        System.out.println(hasCycle(head));

        sc.close();
    }
}